integration-libs/opf/cta/core/services/opf-dynamic-cta.service.ts

```diff
@@ -0,0 +1,252 @@
+
+  protected subList: Array<Subscription> = [];
+  protected isOnsiteMessagingInit = false;
+  protected scriptIdentifiers: Array<string> = [];
+  protected isCartPage: boolean;
+
+  fillCtaRequestforCartPage(
+    scriptLocation: CtaScriptsLocation,
+    paymentAccountIds: number[]
+  ) {
+    this.isCartPage = true;
+    return combineLatest({
+      cart: this.activeCartFacade.takeActive(),
+      additionalData: this.fillAdditionalData(),
+    }).pipe(
+      take(1),
+      map(({ cart, additionalData }) => {
+        return {
+          paymentAccountIds: paymentAccountIds,
+          scriptLocations: [scriptLocation],
+          ctaProductItems: cart.entries?.map((entry) => {
+            return {
+              productId: entry.product?.code,
+              quantity: entry.quantity,
+            } as CtaProductItem;
+          }),
+          additionalData,
+        } as CtaScriptsRequest;
+      })
+    );
+  }
+
+  fillCtaRequestforProductPage(
+    scriptLocation: CtaScriptsLocation,
+    paymentAccountIds: number[]
+  ) {
+    this.isCartPage = false;
+    return combineLatest({
+      product: this.currentProductService.getProduct(),
+      additionalData: this.fillAdditionalData(),
+    }).pipe(
+      take(1),
+      map(({ product, additionalData }) => {
+        return {
+          ctaProductItems: [{ productId: product?.code, quantity: 1 }],
+          paymentAccountIds: paymentAccountIds,
+          scriptLocations: [scriptLocation],
+          additionalData,
+        } as CtaScriptsRequest;
+      })
+    );
+  }
+
+  protected fillAdditionalData(): Observable<Array<OpfKeyValueMap>> {
+    return combineLatest({
+      language: this.languageService.getActive(),
+      currency: this.currencyService.getActive(),
+    }).pipe(
+      take(1),
+      map(({ language, currency }) => {
+        return [
+          {
+            key: 'locale',
+            value: language,
+          },
+          {
+            key: 'currency',
+            value: currency,
+          },
+          {
+            key: 'scriptIdentifier',
+            value: this.getNewScriptIdentifier(),
+          },
+        ];
+      })
+    );
+  }
+
+  initiateEvents() {
+    if (!this.isOnsiteMessagingInit) {
+      this.listenScriptReadyEvent();
+
+      this.isCartPage && this.cartChangedListener();
+      !this.isCartPage && this.productChangedListener();
+
+      this.isOnsiteMessagingInit = true;
+    }
+  }
+
+  stopEvents() {
+    if (this.isOnsiteMessagingInit) {
+      this.subList.forEach((sub) => sub.unsubscribe());
+      this.subList = [];
+      this.globalFunctionsFacade.removeGlobalFunctions(
+        GlobalFunctionsDomain.GLOBAL
+      );
+      this.isOnsiteMessagingInit = false;
+    }
+  }
+
+  protected getNewScriptIdentifier(): string {
+    this.scriptIdentifiers.push(String(this.scriptIdentifiers.length + 1));
+    return String(
+      this.scriptIdentifiers[this.scriptIdentifiers.length - 1]
+    ).padStart(4, '0');
+  }
```

Comment:
[MINOR]
1. I'm having hard time understanding this code.
- Why +1 in one place, but -1 in another place.
- Why wrapping a value from `scriptIdentifiers[i]` within a `String()` again - in line 158? All values inside this array are already strings, arent they?

1. It's not obvious what we need this method for. Let's add JSDocs helpful for newcomer developers reading getting into this method.